core-bundle/src/Resources/contao/dca/tl_page.php

```diff
@@ -590,6 +591,17 @@ static function ($value)
 			'sql'                     => "int(10) unsigned NOT NULL default 0",
 			'relation'                => array('type'=>'hasOne', 'load'=>'lazy')
 		),
+		'subpagesLayout' => array
+		(
+			'exclude'                 => true,
+			'search'                  => true,
+			'inputType'               => 'select',
+			'foreignKey'              => 'tl_layout.name',
+			'options_callback'        => array('tl_page', 'getPageLayouts'),
```

Comment:
This should be a service.

Reply:
But the callback already existed?

Reply:
Ah yes, but without caching the results.

Reply:
Since the existing callback cannot be removed for backwards compatibility, I assume that reusing this callback is preferable to adding a new service that does the same thing.

The caching is added to avoid querying the database twice for the options of both `layout` and `subpagesLayout`.

Both calls could be replaced by a `LayoutOptionsListener`, while the existing `tl_page::getPageOptions` is kept for backwards compatibility. Please let me know if that is what is preferred.

Reply:
> Please let me know if that is what is preferred.

I’d be happy either way. IMO you could integrate the caching into the `getPageLayouts()` method so that we don’t create a new public `loadPageLayouts()` method that needs BC in the future.

Reply:
These are not covered by our BC promise. We have already removed such methods and even whole `tl_` classes in the past.

Reply:
The the `LayoutOptionsListener` would probably be the best option as all core developers seem to be fine with that ☺️